tests/drivers/hwinfo/api/src/main.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2019 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <hwinfo.h>
+#include <ztest.h>
+#include <strings.h>
+
+/*
+ * @addtogroup t_hwinfo_get_device_id_api
+ * @{
+ * @defgroup t_hwinfo_get_device_id test_hwinfo_get_device_id
+ * @brief TestPurpose: verify device id get works
+ * @details
+ * - Test Steps
+ *   -# Read the ID
+ *   -# Check if to many bytes are written to the buffer
+ *   -# Check if UID is plausible
+ * - Expected Results
+ *   -# Device uid with correct length should be written to the buffer.
+ * @}
+ */
+
+#define BUFFER_LENGTH 17
+#define BUFFER_CANARY 0xFF
+
+#if defined(CONFIG_DEVICE_ID_STM32) || defined(CONFIG_DEVICE_ID_NRF)
+
+/*
+ * Function invokes the get_entropy callback in driver
+ * to get the random data and fill to passed buffer
+ */
+static int get_device_id(void)
+{
+	struct device *dev;
+	u8_t buffer[BUFFER_LENGTH];
+	size_t length_read;
+	int i, count;
+
+	dev = device_get_binding(CONFIG_DEVICE_ID_NAME);
+	if (!dev) {
+		TC_PRINT("error: no device id device\n");
+		return TC_FAIL;
+	}
+
+	TC_PRINT("devic id device is %p, name is %s\n",
+		 dev, dev->config->name);
+
+	memset(buffer, BUFFER_CANARY, sizeof(buffer));
+
+	length_read = hwinfo_get_device_id(dev, buffer, BUFFER_LENGTH-1);
+
+	if (buffer[length_read] != BUFFER_CANARY) {
+		TC_PRINT("Fail. Too many bytes are written\n");
+		return TC_FAIL;
+	}
+
+	for (i = 0, count = 0; i < length_read; i++) {
+		if (buffer[i] == BUFFER_CANARY) {
+			count++;
+		}
+	}
+/*
+ * It is not possible to check the end of the written bytes because
+ * they could have the same value as the canary. But if halve of the uid has
+ * the same value there might be something wrong.
+ */
+	if (count > length_read / 2) {
```

Comment:
Rather than guess, you could read it once, copy that into a second buffer, XOR the first `length_read` octets with 0xA5 and the rest with CANARY ^ 0xA5, and read again into that buffer to make sure that what got written matches the original value and the mutated canary isn't changed either.

Reply:
Done.